src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,167 @@
+    private static final int ZERO = 0;
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < ZERO || index > size) {
+            throw new IndexOutOfBoundsException(INDEX_FOR_EXCEPTION
+                    + index + SIZE_FOR_EXCEPTION + size);
+        }
+        Node<T> newNode = new Node<>(value);
+        addNode(newNode, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndexForGet(index);
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndexForGet(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndexForGet(index);
+        Node<T> removedNode = unlinkByIndex(index);
+        return removedNode.value;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T value) {
+        Node<T> current = head;
+        Node<T> previous = null;
+        while (current != null) {
+            if ((value == null && current.value == null)
+                    || (value != null && value.equals(current.value))) {
+                unlinkByValue(previous, current);
+                size--;
+                return true;
+            }
+            previous = current;
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == ZERO;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+
+        Node(T value) {
+            this.value = value;
+        }
+    }
+
+    private void addNode(Node<T> newNode, int index) {
+        if (index == ZERO) {
+            newNode.next = head;
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+        } else if (index == size) {
+            if (tail == null) {
+                head = newNode;
+            } else {
+                tail.next = newNode;
+            }
+            tail = newNode;
+        } else {
+            Node<T> previousNode = getNode(index - NUMBER_TO_SUBTRACT);
+            newNode.next = previousNode.next;
+            previousNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void unlinkByValue(Node<T> previous, Node<T> current) {
```

Comment:
couple points regarding the unlink methods:
* in this one you pass  2 neighboring nodes in params - we need only one
* you have 2 methods, but when it can be only a single `unlink(Node<T> node)` which should just handle all cases if it is tail/mid/head, where you do not need to call `getNode()` or iterate through the list to get that node